### ISSUE TITLE
[aot_eager] [accuracy] [tts_angular] Dtype mismatch

### ISSUE THREAD

**Issue by anijain2305** [pytorch/torchdynamo#1889]
### 🐛 Describe the bug

Accuracy failure - Dtype mismatch

### Minified repro

~~~
from math import inf
import torch
from torch import tensor, device
import torch.fx as fx
import torch._dynamo
from torch._dynamo.testing import rand_strided
from torch._dynamo.debug_utils import run_fwd_maybe_bwd
from torch._dynamo.debug_utils import same_two_models

# REPLACEABLE COMMENT FOR TESTING PURPOSES

args = [((2, 50, 256), (12800, 256, 1), torch.float16, 'cuda', False)]
args = [rand_strided(sh, st, dt, dev).requires_grad_(rg) for (sh, st, dt, dev, rg) in args]


from torch.nn import *
class Repro(torch.nn.Module):
    def __init__(self):
        super().__init__()



    def forward(self, _stack0 : torch.Tensor):
        getitem = _stack0[(slice(None, None, None), -1)];  _stack0 = None
        normalize = torch.nn.functional.normalize(getitem, p = 2, dim = 1);  getitem = None
        return (normalize,)



mod = Repro()
opt_mod = torch._dynamo.optimize("aot_inductor_debug")(mod)
# opt_mod = torch._dynamo.optimize("aot_eager")(mod)


mod.eval()
opt_mod.eval()

class AccuracyError(Exception):
    pass

with torch.cuda.amp.autocast(enabled=True):
    ref = mod(*args)[0]
    res = opt_mod(*args)[0]
    print(f"{ref.dtype}, {res.dtype}")
    assert ref.dtype ==  res.dtype, f"{ref.dtype}, {res.dtype}"





~~~

**Comment by anijain2305** [comment]
The issue seems to be related to `norm` decomp + AMP. Somewhat simplified repro 

~~~

from math import inf
import torch
from torch import tensor, device
import torch.fx as fx
import torch._dynamo
from torch._dynamo.testing import rand_strided
from torch._dynamo.debug_utils import run_fwd_maybe_bwd
from torch._dynamo.debug_utils import same_two_models

# REPLACEABLE COMMENT FOR TESTING PURPOSES

args = [((2, 50, 256), (12800, 256, 1), torch.float16, 'cuda', True)]
args = [rand_strided(sh, st, dt, dev).requires_grad_(rg) for (sh, st, dt, dev, rg) in args]


from torch.nn import *
class Repro(torch.nn.Module):
    def __init__(self):
        super().__init__()



    def forward(self, _stack0 : torch.Tensor):
        getitem = _stack0[(slice(None, None, None), -1)];  _stack0 = None
        normalize = torch.nn.functional.normalize(getitem, p = 2, dim = 1);  getitem = None
        return (normalize,)



mod = Repro()
opt_mod = torch._dynamo.optimize("aot_inductor_debug")(mod)


mod.eval()
opt_mod.eval()

class AccuracyError(Exception):
    pass

with torch.cuda.amp.autocast(enabled=True):
    ref = mod(*args)[0]
    res = opt_mod(*args)[0]
    assert ref.dtype ==  res.dtype



~~~

**Comment by anijain2305** [comment]
This patch resolves the error, but I don't know whats happening here. Thanks @eellison for pointing this out.

~~~
diff --git a/aten/src/ATen/autocast_mode.cpp b/aten/src/ATen/autocast_mode.cpp
index ca75c38258f..7a541f27ddc 100644
--- a/aten/src/ATen/autocast_mode.cpp
+++ b/aten/src/ATen/autocast_mode.cpp
@@ -452,18 +452,18 @@ TORCH_LIBRARY_IMPL(aten, Autocast, m) {
   KERNEL2(cumsum, dimname, fp32_set_opt_dtype)
   // commenting these out because they accept an explicit (not-optional) dtype, and we shouldn't try to flip that even
   // when autocasting.
-  // KERNEL2(norm, ScalarOpt_dtype, fp32_set_opt_dtype)
-  // KERNEL2(norm, ScalarOpt_dim_dtype, fp32_set_opt_dtype)
-  // KERNEL2(norm, names_ScalarOpt_dim_dtype, fp32_set_opt_dtype)
+  KERNEL2(norm, ScalarOpt_dtype, fp32_set_opt_dtype)
+  KERNEL2(norm, ScalarOpt_dim_dtype, fp32_set_opt_dtype)
+  KERNEL2(norm, names_ScalarOpt_dim_dtype, fp32_set_opt_dtype)
   KERNEL(sum, fp32_set_opt_dtype)
   KERNEL2(sum, dim_IntList, fp32_set_opt_dtype)
   KERNEL2(sum, dim_DimnameList, fp32_set_opt_dtype)
   // fp32_append_dtype
   // The fp32_append_dtype wrapper overrides implicit promotion behavior.
   // norm does not implicitly promote, but be aware when adding new ops to this policy.
-  KERNEL_DIFFERENT_REDISPATCH_SIGNATURE(ADD_NS(norm), "norm.Scalar", Tensor (const Tensor &, const Scalar&), Tensor (const Tensor &, const c10::optional<Scalar>&, ScalarType), fp32_append_dtype)
-  KERNEL_DIFFERENT_REDISPATCH_SIGNATURE(ADD_NS(norm), "norm.ScalarOpt_dim", Tensor (const Tensor &, const c10::optional<Scalar>&, IntArrayRef, bool), Tensor (const Tensor &, const c10::optional<Scalar>&, IntArrayRef, bool, ScalarType), fp32_append_dtype)
-  KERNEL_DIFFERENT_REDISPATCH_SIGNATURE(ADD_NS(norm), "norm.names_ScalarOpt_dim", Tensor (const Tensor &, const c10::optional<Scalar>&, DimnameList, bool), Tensor (const Tensor &, const c10::optional<Scalar>&, DimnameList, bool, ScalarType), fp32_append_dtype)
+  // KERNEL_DIFFERENT_REDISPATCH_SIGNATURE(ADD_NS(norm), "norm.Scalar", Tensor (const Tensor &, const Scalar&), Tensor (const Tensor &, const c10::optional<Scalar>&, ScalarType), fp32_append_dtype)
+  // KERNEL_DIFFERENT_REDISPATCH_SIGNATURE(ADD_NS(norm), "norm.ScalarOpt_dim", Tensor (const Tensor &, const c10::optional<Scalar>&, IntArrayRef, bool), Tensor (const Tensor &, const c10::optional<Scalar>&, IntArrayRef, bool, ScalarType), fp32_append_dtype)
+  // KERNEL_DIFFERENT_REDISPATCH_SIGNATURE(ADD_NS(norm), "norm.names_ScalarOpt_dim", Tensor (const Tensor &, const c10::optional<Scalar>&, DimnameList, bool), Tensor (const Tensor &, const c10::optional<Scalar>&, DimnameList, bool, ScalarType), fp32_append_dtype)
   // promote
   KERNEL(addcdiv, promote)
   KERNEL(addcmul, promote)
 ~~~~

**Comment by anijain2305** [comment]
cc @eellison @ezyang

**Comment by anijain2305** [comment]
Somewhat more detail here - https://www.diffchecker.com/xuv0R8HO/. This link shows the diff between the generated graphs on master branch. 

Left - Disabled all the decomps except `norm`. 
Right - Disabled all the decomps even `norm`
